Upgrade agent changes

bernardjkim opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR makes a few improvements to the gravity upgrade agents. 
- The gravity-agents will now run as regular services instead of oneshot services so that they restart on failure and node reboot. 
- A `gravity agent status` command is now available to view the status of the deployed gravity agents. Will look something like:
```
// Successfully received status response from all agents
Tue Jul 14 07:11:19 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.141
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.141
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.141

// Failed to receive status response from all agents
Tue Jul 14 07:12:30 UTC Collecting RPC agent status
Hostname     Address            Status      Version
--------     -------            ------      -------
node-2       172.28.128.102     Offline
node-3       172.28.128.103     Offline
node-1       172.28.128.101     Offline
```

- The `gravity agent deploy` command now supports a `--version` flag to specify a particular agent version to deploy.
- The `gravity agent deploy` command also accepts a node hostname as an argument to deploy an agent on the specified node.
## Type of change
<!--Required. Keep only those that apply.-->

* New feature (non-breaking change which adds functionality)

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/1667
* Requires https://github.com/gravitational/gravity.e/pull/4321

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [x] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->

**Verify gravity-agent is running after reboot**

**Verify lead update agent does not automatically attempt to resume upgrade**

**Verify upgrade can be resumed after node reboot with `gravity upgrade --resume`**

**Verify all gravity-agents are shutdown after upgrade**

**Verify gravity-agents are disabled and do not restart on next reboot**

**Verify upgrade can be resumed after non lead agent restarts**

**Verify upgrade after rollback**

**Deploy particular agent version with --version flag**
```
[vagrant@node-1 installer]$ sudo gravity agent deploy --version=7.1.0-alpha.1.153
Wed Jul 29 19:13:08 UTC Deploying agents on the cluster nodes
Deployed agent on node-3 (172.28.128.103)
Deployed agent on node-1 (172.28.128.101)
Deployed agent on node-2 (172.28.128.102)

[vagrant@node-1 installer]$ sudo gravity agent status
Wed Jul 29 19:13:31 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.153
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.153
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.153
```

**Deploy onto a particular node**
```
[vagrant@node-1 installer]$ sudo gravity agent deploy node-1
Wed Jul 29 19:00:36 UTC Deploying agents on the cluster nodes
Deployed agent on node-1 (172.28.128.101)

[vagrant@node-1 installer]$ sudo gravity agent status
Wed Jul 29 19:00:43 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.154
node-3       172.28.128.103     Offline
node-2       172.28.128.102     Offline
```

**Verify agents show deployed even if older server version**
```
[vagrant@node-1 ~]$ sudo gravity agent status
Wed Jul 29 21:22:44 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     N/A
node-3       172.28.128.103     Deployed     N/A
node-1       172.28.128.101     Deployed     N/A
```

**Verify deploy without args works as usual**
```
[vagrant@node-1 ~]$ sudo gravity agent deploy
Thu Jul 30 21:21:04 UTC Deploying agents on the cluster nodes
Deployed agent on node-2 (172.28.128.102)
Deployed agent on node-3 (172.28.128.103)
Deployed agent on node-1 (172.28.128.101)

[vagrant@node-1 ~]$ sudo gravity agent status
Thu Jul 30 21:21:12 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.162
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.162
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.162
```

bernardjkim committed:
Upgrade agent changes

- Gravity agents are deployed as simple services rather than a oneshot
service.
- Add `gravity agent status` that prints gravity-agent statuses.

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -366,6 +367,30 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
 }
 
+// rpcAgentStatus requests and writes the gravity agent statuses to stdout.
+func rpcAgentStatus(env *localenv.LocalEnvironment) error {
```
How does this method behave / what does it output if there are no deployed agents?

r0mant replied:
Also, what does it output if there's a failure contacting one of the agents / retrieving info from it?

bernardjkim replied:
Added additional details and added a 5 sec timeout on the status request.
If an agent fail to return a status response, the agent will be displayed as `Offline` and the version column will be empty.

r0mant commented on lib/update/cluster/automatic.go:
```diff
@@ -132,3 +132,14 @@ func ShutdownClusterAgents(ctx context.Context, remote rpc.AgentRepository) erro
+
+// AgentStatus requests the statuses of all agents
+func AgentStatus(ctx context.Context, remote rpc.AgentRepository) ([]rpc.GravityAgentStatus, error) {
+	nodes, err := kubectl.GetNodesAddr(ctx)
```
I think we should retrieve the nodes from the cluster state, not k8s. If you pass an operator here, you can use it:

```go
// in cli/rpcagent.go
func rpcAgentStatus(env *localenv.LocalEnvironment) error {
  operator, err := env.ClusterOperator()
  ...
  cluster.AgentStatus(ctx, operator, remote)
}

// here
func AgentStatus(ctx context.Context, operator ops.Operator, remote ...) () {
  cluster, err := operator.GetLocalSite()
  ...
  cluster.ClusterState.Servers // <-- cluster nodes
}
```

r0mant commented on lib/update/cluster/automatic.go:
```diff
@@ -132,3 +132,14 @@ func ShutdownClusterAgents(ctx context.Context, remote rpc.AgentRepository) erro
 }
+
+// AgentStatus requests the statuses of all agents
+func AgentStatus(ctx context.Context, remote rpc.AgentRepository) ([]rpc.GravityAgentStatus, error) {
```
I think this better belongs in `lib/rpc`?

r0mant commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+// GravityAgentStatus contains a gravity agent's status information.
+type GravityAgentStatus struct {
```
Nit: Just `AgentStatus`, or `RPCAgentStatus`.

r0mant commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,118 @@
+	Status string
+	// Version describes gravity agent version.
+	Version string
+}
```
I would add another field here with `Error error` just so it can be displayed in case there's a failure contacting the agent.

r0mant commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,118 @@
+	statusCh := make(chan GravityAgentStatus, len(servers))
+	for _, srv := range servers {
+		go func(host string) {
+			systemInfo, err := getSystemInfo(ctx, host, rpc)
```
I would encapsulate these two (getSystemInfo and getVersion) into a separate function (getAgentStatus) just so the error handling is a little easier, for example:

```go
func getAgentStatus(ctx context.Context, address string, rpc AgentRepository) AgentStatus {
  systemInfo, err := getSystemInfo(ctx, host, rpc)
  // return status with Error set
  version, err := getVersion(ctx, host, rpc)
  // return status with Error set
  return AgentStatus{
    Address:  address,
    Status:   constants.GravityAgentDeployed,
    Hostname: systemInfo.GetHostname(),
    Version:  version.Version,    
  }, nil
}

// Then the goroutine becomes trivial:

go func(address string) {
  statusCh <- getAgentStatus(ctx, address, rpc)
}
```

bernardjkim replied:
Realized we don't even need to request systemInfo since we are only collecting hostname from it. We can pull hostname from the cluster state.

r0mant commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,118 @@
+	// Node specifies the name of the node running the agent.
+	Node string
+	// IP specifies the IP address of the node running the agent.
+	IP string
```
Nit: I'd rename `IP -> Address` and `Node -> Hostname` to be consistent with naming in other places.

r0mant reviewed (commented):
Preliminary feedback for your draft.

bernardjkim committed:
Address feedback

- Use goterm.Table to display status table.
- Un-hide gravity agent subcommands and update desc.
- Update agent deploy to launch simple service.
- Update rpcAgentStatus godoc.
- Retrieve nodes from cluster state.
- Add Error field to AgentStatus.
- IP -> Address; Node -> Hostname.

bernardjkim committed:
Add default WantedBy value

a-palchikov commented on lib/constants/constants.go:
```diff
@@ -230,6 +230,12 @@ const (
 	// OneshotService is a service that executes one time
 	OneshotService = "oneshot"
 
+	// SimpleService is a simple service that is recommended for long running processes
```
I think a link to the systemd docs as a reference to service types would be useful.
Also, I think these could be more appropriate in [lib/system/service](https://github.com/gravitational/gravity/tree/bernard/master/upgrade-agent/lib/system/service) but it's debatable.

a-palchikov commented on lib/rpc/proto/discovery.proto:
```diff
@@ -16,6 +16,15 @@ service Discovery {
+// Version describes an agent's version information
+message Version {
+    // Payload is the version information payload.
+    bytes payload = 1;
```
I would rather have the protobuf to be the source of truth for data types.

bernardjkim replied:
Moved Version definition from modules to proto pkg. a583350

a-palchikov commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,96 @@
+	Status string
+	// Version describes gravity agent version.
+	Version string
+	// Error contains errors while collected while requesting agent status.
```
nit: maybe rephrase this to avoid duplicate `while`: `Error contains an error that might have occurred when requesting agent status`

a-palchikov commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,96 @@
+	Hostname string
+	// Address specifies the IP address of the node running the agent.
+	Address string
+	// Status indiciates the current status of the agent. An agent is `Deployed`
```
indicates

a-palchikov commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func getVersion(ctx context.Context, addr string, rpc AgentRepository) (version modules.Version, err error) {
+	ctx, cancel := context.WithTimeout(ctx, defaults.DialTimeout)
```
This sounds like a bug - `DialTimeout` should only affect the connection step and not the actual API call.

a-palchikov commented on tool/gravity/cli/system.go:
```diff
@@ -280,6 +282,33 @@ func reinstallOneshotService(env *localenv.LocalEnvironment, serviceName string,
 
+// reinstallService stops and reinstalls the service specified by serviceName
+// as a simple service.
+func reinstallService(env *localenv.LocalEnvironment, serviceName string, cmd []string) error {
```
I'd really like that we moved general purpose functions to specific packages - i.e. `lib/system/service` if there's no such API yet.

a-palchikov commented on lib/defaults/defaults.go:
```diff
@@ -219,6 +219,9 @@ const (
 	AgentStopTimeout = 10 * time.Second
 
+	// AgentStatusTimeout specifies the timeout for agent status query
+	AgentStatusTimeout = 5 * time.Second
```
I'm sort of on the fence with timeouts - I feel these should be configurable via envars at least to avoid any issues with slower networks and arbitrary cluster sizes.

bernardjkim replied:
Gravity will now check for `GRAVITY_AGENT_STATUS_TIMEOUT` env var. Will default to 5sec if env var is not specified.

bernardjkim committed:
Move Version definition from modules to proto pkg

bernardjkim committed:
Move system service consts service pkg.

- Move system service config value to service pkg.
- Move ReinstallSimpleService to service pkg.
- Add GetenvDuration func.

bernardjkim committed:
Edit AutomaticUpgrade behavior

AutomaticUpgrade now checks if the upgrade is already in progress. If
so, return without creating a new agent.

bernardjkim committed:
Add Version and Hostname to agent deploy command

- Version can be used to specify the agent version to deploy.
- Hostname can be used to specify on which node to deploy the agent.

bernardjkim committed:
Update gravity.e

r0mant commented on lib/update/cluster/automatic.go:
```diff
@@ -43,6 +43,21 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
+	// Check if the upgrade is already in progress, which would indicate an
+	// agent is already running. Return without creating a new agent.
+	for _, phase := range plan.Phases {
+		if phase.IsInProgress() {
```
I think we should be checking for "not unstarted" phases. Otherwise, this will not catch scenarios where a phase is "failed" or "completed". So we should basically look for phases in any state other than the initial "unstarted" one.

r0mant commented on lib/update/cluster/automatic.go:
```diff
@@ -43,6 +43,21 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
+	// agent is already running. Return without creating a new agent.
+	for _, phase := range plan.Phases {
+		if phase.IsInProgress() {
+			return nil
```
We should definitely log something in this case.

r0mant commented on lib/utils/env.go:
```diff
@@ -131,6 +132,19 @@ func GetenvInt(name string) (int, error) {
 	return valueI, nil
 }
 
+//GetenvDuration returns the specified environment variable value parsed as a duration.
```
Nit: Space after //.

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -315,16 +331,44 @@ func deployAgents(ctx context.Context, env *localenv.LocalEnvironment, req deplo
+
+	// If version is not specified in the request, use the current build version
+	if req.version == "" {
+		req.version = strings.Split(version.Get().Version, "+")[0]
```
What's "split by +" for?

bernardjkim replied:
https://github.com/gravitational/gravity/blob/46b9ee065b121ed99ac593cb98084ef079ac020b/tool/gravity/cli/rpcagent.go#L416
Copied over from `getGravityPackage` func. It doesn't seem like we need to split version by '+'. I guess we had a different version format in the past?

r0mant replied:
@a-palchikov Do you recall by any chance why we had this "split by +"? I remember it was needed for something but can't remember why.

a-palchikov replied:
It might have had something to do with semver parsing but I need to go through the history to know more/remember.

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
+			log.WithError(err).Debugf("Failed to collect query agent version on %s.", status.Address)
+		}
+	}
+	fmt.Println(t.String())
```
We usually use `env.Println` - it considers some flags potentially passed on the CLI, like "quiet" or "verbose" that might affect the output.

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
+	for _, status := range statusList {
+		fmt.Fprintf(t, "%s\t%s\t%s\t%s\n", status.Hostname, status.Address, status.Status, status.Version)
+		if status.Error != nil {
+			log.WithError(err).Debugf("Failed to collect query agent version on %s.", status.Address)
```
Shouldn't this be `log.WithError(status.Error)`?

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
+	}
+	fmt.Println(t.String())
+
+	return trace.Wrap(err)
```
It's a bit weird - what is this error that's returned here? The duration parsing error from above?

r0mant commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -345,6 +389,17 @@ func newDeployAgentsRequest(ctx context.Context, env *localenv.LocalEnvironment,
 
+// findServerByHostname returns the server with the matching hostname. Returns
+// a NotFound error if a server with a matching hostname does not exist.
+func findServerByHostname(clusterState storage.ClusterState, hostname string) (server storage.Server, err error) {
```
I'm pretty sure we already have this function in storage.go.

https://github.com/gravitational/gravity/blob/master/lib/storage/storage.go#L696

r0mant commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	version, err := getVersion(ctx, server.AdvertiseIP, rpc)
+	if err != nil {
```
I think this should also consider a "not implemented" error, in case it contacts the agent that doesn't have GetVersion method yet. In this case the agent is basically online, we just don't know its version (could display it as `N/A` for example).

bernardjkim committed:
Address feedback

- Report 'N/A' version if rpc NotImplemented.
- Check for phase is !IsUnstarted.
- Remove duplicate findServerByHostname func.
- No need to split version by "+".
- Fix error handling.
- Use env.Println.

a-palchikov commented on lib/rpc/status.go:
```diff
@@ -0,0 +1,105 @@
+	for range servers {
+		statusList = append(statusList, <-statusCh)
+	}
+
```
nit: could also close the channel at this point to help gc.

a-palchikov commented on lib/update/cluster/automatic.go:
```diff
@@ -43,6 +43,22 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
+
+	// Check if the upgrade is already in progress, which would indicate an
+	// agent is already running. Return without creating a new agent.
+	for _, phase := range plan.Phases {
```
This should be better encapsulated as `IsOperationInProgress` or `IsPlanInProgress` - you would not need the comment then. I'd check whether we might already have an API like that.

a-palchikov commented on tool/gravity/cli/checks.go:
```diff
@@ -134,7 +134,7 @@ func checkUpgrade(ctx context.Context, env *localenv.LocalEnvironment, config pr
 	}
 	// Deploy RPC agents that will be used for running checks on the nodes.
-	credentials, err := rpcAgentDeployHelper(ctx, env, "", "")
+	credentials, err := rpcAgentDeployHelper(ctx, env, "", "", "", "")
```
Pretty hard to tell what the values are and why empty. In this case I'd suggest an options struct which would be provided to the function as empty.

a-palchikov commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -388,6 +432,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
+	for _, status := range statusList {
+		fmt.Fprintf(t, "%s\t%s\t%s\t%s\n", status.Hostname, status.Address, status.Status, status.Version)
+		if status.Error != nil {
+			log.WithError(status.Error).Debugf("Failed to collect agent status on %s.", status.Address)
```
I wonder if we need to indicate with that command whether (any) agents are offline via a non-0 exit code?
Also, if used with `--quiet` the command will not output anything - not sure if this is intentional/desired.

bernardjkim replied:
It could be useful to have the command return a non-0 exit code if any agents are offline. Updated in cd7ee63.

I think it was preferred to use env.Println. Not entirely sure when someone would want to run the agent status command without any output, but since it now returns an error if any agents are offline, it might be useful somewhere.

a-palchikov commented on tool/gravity/cli/rpcagent.go:
```diff
@@ -315,16 +331,44 @@ func deployAgents(ctx context.Context, env *localenv.LocalEnvironment, req deplo
-		return nil, trace.Wrap(err)
+	// If hostname is specified in the request, deploy agent only on specified node
+	var servers []rpc.DeployServer
+	if req.hostname != "" {
```
Not sure about the need to leak hostname into deployAgentRequest - if it's desired that only an agent on a specific node is started, add an optional server (or a server slice - a single element means only on a single node) to the request.

bernardjkim committed:
Address feedback

- Close channel.
- Add IsStarted func.
- Add deployOptions struct.
- Return non-0 if any agent is offline.
- Replace hostname with servers from deployAgentsRequest.

bernardjkim committed:
Edit error msg

bernardjkim committed:
Convert grpc to trace

r0mant commented on lib/system/service/simple.go:
```diff
@@ -0,0 +1,55 @@
+
+	err = services.StopService(serviceName)
+	if err != nil {
+		logrus.Warnf("Error stopping service %v: %v.", serviceName, err)
```
`logrus.WithError(err).Warnf()`

r0mant reviewed (approved):
Code looks good, one additional suggestion for "testing done" is in addition to `Verify all gravity-agents are shutdown after upgrade` scenario, let's also make sure that the shutdown agents are not automatically restarted after reboot (i.e. complete the upgrade, reboot the node(-s) and make sure agents aren't running).

bernardjkim committed:
upgrade and gc need to specify servers to deployAgentsRequest

bernardjkim committed:
Specify version in upgrade and gc deployAgentsRequest

bernardjkim committed:
Disable service after run completes

bernardjkim committed:
Disable agent service on shutdown

bernardjkim committed:
Update ShutdownAgent rpc to disable service before shutdown

bernardjkim committed:
Warnf

bernardjkim committed:
Update gravity.e

bernardjkim committed:
OneshotService moved to service pkg
